auditlog/log-enricher/src/main/java/io/debezium/demos/auditing/enricher/TopologyProducer.java

```diff
@@ -36,73 +46,235 @@
     public Topology buildTopology() {
         StreamsBuilder builder = new StreamsBuilder();
 
+        StoreBuilder<KeyValueStore<Long, JsonObject>> streamBufferStateStore =
+                Stores
+                    .keyValueStoreBuilder(
+                        Stores.persistentKeyValueStore(STREAM_BUFFER_NAME),
+                        new Serdes.LongSerde(),
+                        new JsonObjectSerde()
+                    )
+                    .withCachingDisabled();
+            builder.addStateStore(streamBufferStateStore);
+
         builder.globalTable(txContextDataTopic, Materialized.as(STORE_NAME));
 
         builder.<JsonObject, JsonObject>stream(vegetablesTopic)
                 // filter out any tombstones; for an audit log topic, time-based retention
                 // seems more reasonable than compaction
                 .filter((id, changeEvent) -> changeEvent != null)
-
-                // enrich change events with transaction metadata via the statestore of the TX
-                // topic
-                .transform(() -> new ChangeEventEnricher())
+                // exclude snapshot events
+                .filter((id, changeEvent) -> !changeEvent.getString("op").equals("r"))
+                // enrich change events with transaction metadata via the statestore of the TX topic
+                .transform(() -> new ChangeEventEnricher(), STREAM_BUFFER_NAME)
                 .to(vegetablesEnrichedTopic);
 
         return builder.build();
     }
 
+    /**
+     * Enriches change events with transaction-scoped metadata. If no metadata for
+     * the associated transaction can be retrieved yet (as the change event gets
+     * processed before the corresponding transaction record), that change event
+     * gets added into a buffer. Before processing the incoming change event, any
+     * buffered events will be processed.
+     */
     private final class ChangeEventEnricher implements Transformer<JsonObject, JsonObject, KeyValue<JsonObject, JsonObject>> {
 
+        private ProcessorContext context;
         private KeyValueStore<JsonObject, JsonObject> txMetaDataStore;
+        private KeyValueStore<Long, JsonObject> streamBuffer;
 
         @Override
         @SuppressWarnings("unchecked")
         public void init(ProcessorContext context) {
+            this.context = context;
+            streamBuffer = (KeyValueStore<Long, JsonObject>) context.getStateStore(STREAM_BUFFER_NAME);
             txMetaDataStore = (KeyValueStore<JsonObject, JsonObject>) context.getStateStore(STORE_NAME);
+
+            context.schedule(Duration.ofSeconds(1), PunctuationType.WALL_CLOCK_TIME, ts -> {
```

Comment:
nitpick: no need for block lambda ;-)

Reply:
Yeah, that was a left-over from before where I had some logging code. Removed.